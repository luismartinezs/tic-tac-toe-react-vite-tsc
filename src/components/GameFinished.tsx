import { useState } from "react";
import calculateWinner from "@/utils/calculateWinner";

const Square = ({ value, onClick }) => {
  return (
    <button className="square" onClick={() => onClick()}>
      {value}
    </button>
  );
};

const Board = ({ squares, onClick }) => {
  const renderSquare = (i) => (
    <Square value={squares[i]} onClick={() => onClick(i)} />
  );

  return (
    <div>
      <div className="board-row">
        {renderSquare(0)}
        {renderSquare(1)}
        {renderSquare(2)}
      </div>
      <div className="board-row">
        {renderSquare(3)}
        {renderSquare(4)}
        {renderSquare(5)}
      </div>
      <div className="board-row">
        {renderSquare(6)}
        {renderSquare(7)}
        {renderSquare(8)}
      </div>
    </div>
  );
};

const Game = () => {
  const [history, setHistory] = useState([
    {
      squares: Array(9).fill(null),
    },
  ]);

  const [isXNext, setIsXNext] = useState(true);
  const [stepNumber, setStepNumber] = useState(0);

  const current = history[stepNumber];

  const handleClick = (i) => {
    const _history = history.slice(0, stepNumber + 1);
    const current = _history.at(-1);
    const squares = [...current.squares];

    if (calculateWinner(squares) || squares[i]) {
      return;
    }

    squares[i] = isXNext ? "X" : "O";

    const newHistory = _history.concat({ squares })
    setHistory(newHistory);
    setStepNumber(newHistory.length);
    setIsXNext(!isXNext);
  };

  const winner = calculateWinner(current.squares);
  let status;
  if (winner) {
    status = `Winner: ${winner}`;
  } else {
    status = `Next player: ${isXNext ? "X" : "O"}`;
  }

  const jumpTo = (step) => {
    setStepNumber(step);
    setIsXNext(step % 2 === 0);
  };

  const moves = history.map((step, move) => {
    const desc = move ? `Go to move # ${move}` : "Go to game start";
    return (
      <li key={move}>
        <button className={move === stepNumber ? 'select-move active' : 'select-move'} onClick={() => jumpTo(move)}>{desc}</button>
      </li>
    );
  });

  return (
    <div className="game">
      <div className="game-board">
        <Board squares={current.squares} onClick={(i) => handleClick(i)} />
      </div>
      <div className="game-info">
        <div className="status">{status}</div>
        <ol>{moves}</ol>
      </div>
    </div>
  );
};

export default Game;
