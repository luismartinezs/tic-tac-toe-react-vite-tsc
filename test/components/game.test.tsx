import React from "react";
import { Square, Board, Game } from "@/components/GameFinished.tsx";
import { describe, it, expect } from "vitest";
import renderer, { act, create } from "react-test-renderer";

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON();
  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);
  return result as renderer.ReactTestRendererJSON;
}

describe("Square", () => {
  let component;
  const testValue = (value) => {
    act(() => {
      if (component) {
        // update component
        component.update(<Square value={value} onClick={() => null} />);
      } else {
        // initialize component
        component = create(<Square value={value} onClick={() => null} />);
      }
    });
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  };
  it("Renders a button with a label dependent on props.value", () => {
    testValue("X");
    testValue("O");
    testValue(null);
  });
});

describe("Board", () => {
  it("Renders board", () => {
    const squares = Array(9).fill(null);
    const component = create(<Board squares={squares} onClick={() => null} />);
    // console.log(component.root.findByType(Square))
    // console.log(component.root.findAllByType(Square)) // all Square children
    // console.log(component.root.props);
    // console.log(component.root.children);
    let tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});

describe("Game", () => {
  let component
  act(() => {
    component = create(<Game />);
  })
  let tree = toJson(component);
  let squares = component.root.findAllByType(Square)
  console.log(squares[0]._fiber.child)

  it("Renders game", () => {
    expect(tree).toMatchSnapshot();
  });
  it("Clicking on square fills it with X", () => {

  });
  it.todo("Clicking on square switches status to 'Next player: O'");
  it.todo("Clicking on square adds new item to moves list");
  it.todo("Clicking on first item of moves list resets Board state");
  it.todo(
    "Clicking on last item of moves list returns Board state to that move"
  );
  it.todo("Clicking on any one moves list item, resets Board to that move");
  it.todo("When X wins, show 'Winner: X' message");
  it.todo("After a player wins, clicking on a square does nothing");
  it.todo("Clicking on Show finish / start button toggles the game");
});
