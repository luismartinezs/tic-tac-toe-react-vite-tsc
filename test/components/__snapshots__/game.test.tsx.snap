// Vitest Snapshot v1

exports[`Board > Renders board 1`] = `
<div>
  <div
    className="board-row"
  >
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>
  <div
    className="board-row"
  >
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>
  <div
    className="board-row"
  >
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`Game > Renders game 1`] = `
<div
  className="game"
>
  <div
    className="game-board"
  >
    <div>
      <div
        className="board-row"
      >
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
      </div>
      <div
        className="board-row"
      >
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
      </div>
      <div
        className="board-row"
      >
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
      </div>
    </div>
  </div>
  <div
    className="game-info"
  >
    <div
      className="status"
    >
      Next player: X
    </div>
    <ol>
      <li>
        <button
          className="select-move active"
          onClick={[Function]}
        >
          Go to game start
        </button>
      </li>
    </ol>
  </div>
</div>
`;

exports[`Square > Renders a button 1`] = `
<button
  className="square"
  onClick={[Function]}
>
  X
</button>
`;

exports[`Square > Renders a button 2`] = `
<button
  className="square"
  onClick={[Function]}
>
  O
</button>
`;

exports[`Square > Renders a button 3`] = `
<button
  className="square"
  onClick={[Function]}
/>
`;

exports[`Square > Renders a button with a label dependent on props.value 1`] = `
<button
  className="square"
  onClick={[Function]}
>
  X
</button>
`;

exports[`Square > Renders a button with a label dependent on props.value 2`] = `
<button
  className="square"
  onClick={[Function]}
>
  O
</button>
`;

exports[`Square > Renders a button with a label dependent on props.value 3`] = `
<button
  className="square"
  onClick={[Function]}
/>
`;
